_delphi_utils_python/delphi_utils/smooth.py

```diff
@@ -0,0 +1,434 @@
+            np.convolve(
+                signal_padded, np.ones(self.window_length, dtype=int), mode="valid"
+            )
+            / self.window_length
+        )
+
+        return signal_smoothed
+
+    def left_gauss_linear_smoother(self, signal):
+        """
+        DEPRECATED: This method is available to help sanity check the 'savgol' method.
+        It is a little slow, so use 'savgol' with poly_fit_degree=1 instead.
+
+        Smooth the y-values using a local linear regression with Gaussian weights.
+
+        At each time t, we use the data from times 1, ..., t-dt, weighted
+        using the Gaussian kernel, to produce the estimate at time t.
+
+        For math details, see the smoothing_docs folder.
+
+        Parameters
+        ----------
+        signal: np.ndarray
+            A 1D signal.
+
+        Returns
+        ----------
+        signal_smoothed: np.ndarray
+            A smoothed 1D signal.
+        """
+        warnings.warn(
+            "Use the savgol smoother with poly_fit_degree=1 instead.",
+            DeprecationWarning,
+        )
+        n = len(signal)
+        signal_smoothed = np.zeros_like(signal)
+        A = np.vstack([np.ones(n), np.arange(n)]).T  # the regression design matrix
+        for idx in range(n):
+            weights = np.exp(
+                -((np.arange(idx + 1) - idx) ** 2) / self.gaussian_bandwidth
+            )
+            AwA = np.dot(A[: (idx + 1), :].T * weights, A[: (idx + 1), :])
+            Awy = np.dot(
+                A[: (idx + 1), :].T * weights, signal[: (idx + 1)].reshape(-1, 1)
+            )
+            try:
+                beta = np.linalg.solve(AwA, Awy)
+                signal_smoothed[idx] = np.dot(A[: (idx + 1), :], beta)[-1]
+            except np.linalg.LinAlgError:
+                signal_smoothed[idx] = signal[idx] if self.impute else np.nan
+        if self.minval is not None:
+            signal_smoothed[signal_smoothed <= self.minval] = self.minval
+        return signal_smoothed
+
+    def savgol_predict(self, signal):
+        """
+        Fits a polynomial through the values given by the signal and returns the value
+        of the polynomial at the right-most signal-value. More precisely, fits a polynomial
+        f(t) of degree poly_fit_degree through the points signal[0], signal[1] ..., signal[-1],
+        and returns the evaluation of the polynomial at the location of signal[-1].
+
+        Parameters
+        ----------
+        signal: np.ndarray
+            A 1D signal to smooth.
+
+        Returns
+        ----------
+        coeffs: np.ndarray
+            A vector of coefficients of length nl that determines the savgol
+            convolution filter.
+        """
+        coeffs = self.savgol_coeffs(
+            -len(signal) + 1, 0, self.poly_fit_degree, self.gaussian_bandwidth
+        )
+        return signal @ coeffs
+
+    @classmethod
+    def savgol_coeffs(cls, nl, nr, poly_fit_degree, gaussian_bandwidth=100):
```

Comment:
I can't see that this function is ever used as a class method (although I'm not familiar with this functionality).

Reply:
Thanks for catching the different default. I think the `@classmethod` is left over from when I was testing this class and wanted to call this function without instantiating an object. At this point, it's probably unlikely anyone else will call it independently, so we can probably remove that.